common/crypto/CMakeLists.txt

```diff
@@ -0,0 +1,94 @@
+ENDIF()
+
+FILE(GLOB PROJECT_HEADERS *.h verify_ias_report/*.h)
+FILE(GLOB PROJECT_SOURCES *.cpp verify_ias_report/*.cpp)
+
+################################################################################
+# Generated source
+################################################################################
+
+# this is a dependency to generate the `ias-certificates.cpp` file
+SET(PROJECT_GENERATED_IAS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/verify_ias_report/ias-certificates.cpp)
+SET_SOURCE_FILES_PROPERTIES(${PROJECT_GENERATED_IAS_SOURCES} PROPERTIES GENERATED TRUE)
+ADD_CUSTOM_COMMAND(OUTPUT  ${PROJECT_GENERATED_IAS_SOURCES}
+                     COMMAND ./build_ias_certificates_cpp.sh
+                     DEPENDS verify_ias_report/ias-certificates.template verify_ias_report/build_ias_certificates_cpp.sh
+                     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/verify_ias_report
+)
+ADD_CUSTOM_TARGET(generate-ias-files DEPENDS ${PROJECT_GENERATED_IAS_SOURCES})
+
+################################################################################
+# Common crypto settings
+################################################################################
+
+IF("$ENV{SGX_MODE}" STREQUAL "SIM")
+    SET(IAS_CA_CERT_REQUIRED_FLAGS "")
+ELSE()
+    SET(IAS_CA_CERT_REQUIRED_FLAGS "-DIAS_CA_CERT_REQUIRED")
+ENDIF()
+
+SET(COMMON_CXX_FLAGS ${IAS_CA_CERT_REQUIRED_FLAGS} "-m64" "-fvisibility=hidden" "-fpie" "-fPIC" "-fstack-protector" "-std=c++11" "-Wall")
+
+# required for: `types.h` and `error.h` in common
+SET(PRIVATE_DEPS_INCLUDE_DIRS "${CMAKE_CURRENT_SOURCE_DIR}/.." "${CMAKE_CURRENT_SOURCE_DIR}/../packages/base64")
+
+# required for:`sgx_error.h` in sgx-sdk
+SET(PUBLIC_INCLUDE_DIRS "$ENV{SGX_SDK}/include")
+
+################################################################################
+# Trusted Crypto library
+################################################################################
+
+SET(T_INCLUDE_DIRS ${PUBLIC_INCLUDE_DIRS})
+# required for: `tSgxSSL_api.h` in sgx-ssl
+LIST(APPEND T_INCLUDE_DIRS "$ENV{SGX_SSL}/include")
+# required for stdlibc++
+LIST(APPEND T_INCLUDE_DIRS "$ENV{SGX_SDK}/include/libcxx")
+# required for `sys/endian.h`
+LIST(APPEND T_INCLUDE_DIRS "$ENV{SGX_SDK}/include/tlibc")
+
+ADD_LIBRARY(${T_CRYPTO_LIB_NAME} STATIC ${PROJECT_HEADERS} ${PROJECT_GENERATED_IAS_SOURCES} ${PROJECT_SOURCES})
+ADD_DEPENDENCIES(${T_CRYPTO_LIB_NAME} generate-ias-files)
+
+TARGET_INCLUDE_DIRECTORIES(${T_CRYPTO_LIB_NAME} PRIVATE ${PRIVATE_DEPS_INCLUDE_DIRS})
+TARGET_COMPILE_OPTIONS(${T_CRYPTO_LIB_NAME} PRIVATE ${COMMON_CXX_FLAGS})
+TARGET_COMPILE_OPTIONS(${T_CRYPTO_LIB_NAME} PRIVATE -nostdinc)
+TARGET_INCLUDE_DIRECTORIES(${T_CRYPTO_LIB_NAME} PRIVATE ${T_INCLUDE_DIRS})
+
+################################################################################
+# Untrusted Crypto library
+################################################################################
+
+SET(U_INCLUDE_DIRS ${PUBLIC_INCLUDE_DIRS})
```

Comment:
This comes back to my question yesterday about being able to use `updo-crypto` without the SGX SDK (say, in Graphene). The reason why we need to include `sgx_error.h` in `updo-crypto` is because of `error.h`, though as far as I can tell, there isn't any untrusted PDO code that actually uses `SgxError`, is there? Does FPC need this? If we would place `#ifndef _UNTRUSTED_` around those declarations in `error.h`, we could remove the dependency on the SDK for `updo-crypto`.

Reply:
> This comes back to my question yesterday about being able to use `updo-crypto` without the SGX SDK (say, in Graphene). The reason why we need to include `sgx_error.h` in `updo-crypto` is because of `error.h`, though as far as I can tell, there isn't any untrusted PDO code that actually uses `SgxError`, is there? 

Right, there shouldn't be any -- I have not checked that.

> Does FPC need this?

No.

> If we would place `#ifndef _UNTRUSTED_` around those declarations in `error.h`, we could remove the dependency on the SDK for `updo-crypto`, which I think would be nice to not have.

I agree, and if this helps with Graphene, I would be happy to review and approve the modifications.

Reply:
I'll open an issue for this.